src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al

```diff
@@ -0,0 +1,518 @@
+                CreateNoSeries(NoSeriesProposal, NoSeriesGenerated, Completion);
+            end else
+                ResponseText := Completion;
+        end;
+    end;
+
+    procedure ApplyProposedNoSeries(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    begin
+        if NoSeriesGenerated.FindSet() then
+            repeat
+                InsertNoSeriesWithLines(NoSeriesGenerated);
+                ApplyNoSeriesToSetup(NoSeriesGenerated);
+            until NoSeriesGenerated.Next() = 0;
+    end;
+
+    local procedure InsertNoSeriesWithLines(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    begin
+        InsertNoSeries(NoSeriesGenerated);
+        InsertNoSeriesLine(NoSeriesGenerated);
+    end;
+
+    local procedure InsertNoSeries(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    var
+        NoSeries: Record "No. Series";
+    begin
+        NoSeries.Init();
+        NoSeries.Code := NoSeriesGenerated."Series Code";
+        NoSeries.Description := NoSeriesGenerated.Description;
+        NoSeries."Manual Nos." := true;
+        NoSeries."Default Nos." := true;
+        //TODO: Check if we need to add more fields here, like "Mask", "No. Series Type", "Reverse Sales VAT No. Series" etc.
+        if not NoSeries.Insert(true) then
+            NoSeries.Modify(true);
+    end;
+
+    local procedure InsertNoSeriesLine(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    var
+        NoSeriesLine: Record "No. Series Line";
+    begin
+        NoSeriesLine.Init();
+        NoSeriesLine."Series Code" := NoSeriesGenerated."Series Code";
+        NoSeriesLine."Line No." := GetNoSeriesLineNo(NoSeriesGenerated."Series Code");
+        NoSeriesLine."Starting No." := NoSeriesGenerated."Starting No.";
+        NoSeriesLine."Ending No." := NoSeriesGenerated."Ending No.";
+        NoSeriesLine."Warning No." := NoSeriesGenerated."Warning No.";
+        NoSeriesLine."Increment-by No." := NoSeriesGenerated."Increment-by No.";
+        //TODO: Check if we need to add more fields here, like "Allow Gaps in Nos.", "Sequence Name" etc.
+        if not NoSeriesLine.Insert(true) then
+            NoSeriesLine.Modify(true);
+    end;
+
+    local procedure GetNoSeriesLineNo(SeriesCode: Code[20]): Integer
+    var
+        NoSeriesLine: Record "No. Series Line";
+        NoSeries: Codeunit "No. Series";
+    begin
+        if not NoSeries.GetNoSeriesLine(NoSeriesLine, SeriesCode, 0D, true) then
+            exit(1000);
+
+        exit(NoSeriesLine."Line No."); // TODO: Check if we need to update existing no series line, or add a new one, e.g. if user requested to create no. series for the new year
+    end;
+
+    local procedure ApplyNoSeriesToSetup(var NoSeriesGenerated: Record "No. Series Proposal Line")
+    var
+        RecRef: RecordRef;
+        FieldRef: FieldRef;
+    begin
+        RecRef.Open(NoSeriesGenerated."Setup Table No.");
+        if not RecRef.FindFirst() then
```

Comment:
Since the assumption here seems to be that there is only 1 setup record and that record would have a blank primary key,  should we not assert that more aggressively? Instead of finding the first record maybe we should 'get' the record with the blank prim. key. ... or check that the record we find has a blank value.